op/model.py

```diff
@@ -186,8 +188,17 @@ def __repr__(self):
 class RelationData(Mapping):
     def __init__(self, relation, our_unit, backend):
         self.relation = weakref.proxy(relation)
-        self._data = {our_unit: RelationUnitData(self.relation, our_unit, True, backend)}
-        self._data.update({unit: RelationUnitData(self.relation, unit, False, backend) for unit in self.relation.units})
+        self._data = {our_unit: RelationDataContent(self.relation, our_unit, (lambda _: True), backend)}
+
+        # Whether the application data bag is mutable or not depends on whether this unit is a leader or not,
+        # but this is not guaranteed to be always true during the same hook execution.
+        self._data.update({our_unit.app: RelationDataContent(self.relation, our_unit.app, (lambda backend: backend.is_leader()), backend)})
```

Comment:
Same.. this doesn't look clean at all. Note the hint of how the documentation is talking about the details which are not relevant in this place. RelationData shouldn't care about such things in its constructor ideally.

Reply:
+1, answered in https://github.com/canonical/operator/pull/47#discussion_r352296886